druid/src/widget/tabs.rs

```diff
@@ -0,0 +1,960 @@
+    }
+
+    pub fn live(&self) -> bool {
+        self.current_time < self.duration
+    }
+
+    pub fn fraction(&self) -> f64 {
+        (self.current_time as f64) / (self.duration as f64)
+    }
+
+    pub fn previous_transform(&self, axis: Axis, main: f64) -> Affine {
+        let x = if self.increasing {
+            -main * self.fraction()
+        } else {
+            main * self.fraction()
+        };
+        Affine::translate(axis.pack(x, 0.))
+    }
+
+    pub fn selected_transform(&self, axis: Axis, main: f64) -> Affine {
+        let x = if self.increasing {
+            main * (1.0 - self.fraction())
+        } else {
+            -main * (1.0 - self.fraction())
+        };
+        Affine::translate(axis.pack(x, 0.))
+    }
+}
+
+fn ensure_for_tabs<Content, TP: TabsPolicy + ?Sized>(
+    contents: &mut Vec<(TP::Key, Content)>,
+    policy: &TP,
+    data: &TP::Input,
+    f: impl Fn(&TP, TP::Key) -> Content,
+) -> Vec<usize> {
+    let mut existing_by_key: HashMap<TP::Key, Content> = contents.drain(..).collect();
+
+    let mut existing_idx = Vec::new();
+    for key in policy.tabs(data).into_iter() {
+        let next = if let Some(child) = existing_by_key.remove(&key) {
+            existing_idx.push(contents.len());
+            child
+        } else {
+            f(&policy, key.clone())
+        };
+        contents.push((key.clone(), next))
+    }
+    existing_idx
+}
+
+pub struct TabsBody<TP: TabsPolicy> {
+    children: Vec<(TP::Key, Option<TabBodyPod<TP>>)>,
+    axis: Axis,
+    transition: TabsTransition,
+    transition_state: Option<TabsTransitionState>,
+    phantom_tp: PhantomData<TP>,
+}
+
+impl<TP: TabsPolicy> TabsBody<TP> {
+    pub fn new(axis: Axis, transition: TabsTransition) -> TabsBody<TP> {
+        TabsBody {
+            children: vec![],
+            axis,
+            transition,
+            transition_state: None,
+            phantom_tp: Default::default(),
+        }
+    }
+
+    fn make_tabs(&mut self, data: &TabsState<TP>) -> Vec<usize> {
+        ensure_for_tabs(
+            &mut self.children,
+            &data.policy,
+            &data.inner,
+            |policy, key| policy.tab_body(key, &data.inner).map(WidgetPod::new),
+        )
+    }
+
+    fn active_child(&mut self, state: &TabsState<TP>) -> Option<&mut TabBodyPod<TP>> {
+        Self::child(&mut self.children, state.selected)
+    }
+
+    // Doesn't take self to allow separate borrowing
+    fn child(
+        children: &mut Vec<(TP::Key, Option<TabBodyPod<TP>>)>,
+        idx: usize,
+    ) -> Option<&mut TabBodyPod<TP>> {
+        children.get_mut(idx).and_then(|x| x.1.as_mut())
+    }
+
+    fn child_pods(&mut self) -> impl Iterator<Item = &mut TabBodyPod<TP>> {
+        self.children.iter_mut().flat_map(|x| x.1.as_mut())
+    }
+}
+
+fn hidden_should_receive_event(evt: &Event) -> bool {
+    match evt {
+        Event::WindowConnected
+        | Event::WindowSize(_)
+        | Event::Timer(_)
+        | Event::Command(_)
```

Comment:
Just a comment, that if I close the tabs example before clicking on the various tabs, I get a bunch of warnings like
```
WARN  [druid::core] Widget 'druid::widget::label::Label<tabs::Advanced>' received an event (Command(Command { symbol: "druid-builtin.close-window", payload: Any })) without having been laid out. This likely indicates a missed call to set_layout_rect.
```

My inclination is that everything is fine, and the warning is overzealous. But at some point it's probably worth thinking (and documenting) what should happen in general for hidden widgets.

Reply:
This might be that I need more detailed info on the events, ie which ones should be delivered to widgets that haven't been seen.

Reply:
I've changed this to layout all the children now, even if not active or animating, which I think resolves this.

Reply:
I think that's reasonable for now. I agree at some point we need to think about how to deal with hiding and unhiding widgets, but I don't mind if we're just doing a bit of extra work for the time being.